Terra Cannot Import Inlined C Functions

As I discovered while trying to get the tests to work again in #400, Terra shouldn't actually be working on windows _at all_. This is because `printf` is defined as an inline function and [has been since Visual Studio 2015](https://social.msdn.microsoft.com/Forums/vstudio/en-US/5150eeec-4427-440f-ab19-aecb26113d31/updated-to-vs-2015-and-now-get-unresolved-external-errors?forum=vcgeneral). As a result, `printf` doesn't exist. You can force it to exist by including `legacy_stdio_definitions.lib`, but you must include this manually on _every single project_ on windows that uses `printf`, without disrupting linux compilations, like so:
```
terralib.saveobj("benchmark_fannkuchredux", { main = main }, (jit.os == "Windows" and {"\\legacy_stdio_definitions.lib"} or nil) )
```
The larger issue here is that terra cannot actually compile any inline-only function because the function immediately ceases to exist. This also applies to things like compiler intrinsics that don't correspond to an actual function symbol. 

The inline-only problem can be fixed by using `llvm.used`, as was done in #405, but this does not fix the stdio specific symbol resolution problem.

We could also consider trying to just manually include the `stdio_legacy_definitions.lib` on all windows compilations (plus our own library, if necessary) and therefore ensure that all of the _standard library functions_ have symbols, at the very least. Whatever option is chosen, using the C standard library's `printf()` function should not result in random, confusing compiler errors that are incredibly hard to track down.

The inline function issue has been resolved using `llvm.used`, but there is still a windows-specific `stdio.h` problem with duplicate symbols:
```
msvcrt.lib(default_local_stdio_options.obj) : error LNK2005: __local_stdio_printf_options already defined in terra-19860a.o
stdio.exe : fatal error LNK1169: one or more multiply defined symbols found
```

This function has the following definition:
```
// This function must not be inlined into callers to avoid ODR violations.  The
// static local variable has different names in C and in C++ translation units.
_Check_return_ _Ret_notnull_
__declspec(noinline) __inline unsigned __int64* __CRTDECL __local_stdio_printf_options(void)
{
    static unsigned __int64 _OptionsStorage;
    return &_OptionsStorage;
}
```

Certain cases that already exist in terra (usually ones with lots of other C code involved) can use `stdio.h` without causing this problem. However, a minimal case, where only `stdio.h` is included, fails. Clang itself does not run into this problem, but seems to declare this symbol as ExternallyAvailable, which is not what `stdio.h` actually does:
```
$__local_stdio_printf_options = comdat any

@"??_C@_06FJFFADID@C?3?5?$CFi?6?$AA@" = linkonce_odr dso_local unnamed_addr constant [7 x i8] c"C: %i\0A\00", comdat, align 1
@__local_stdio_printf_options._OptionsStorage = internal global i64 0, align
```
The LLVM that Terra gets back from clang, however, is an actual function definition: 
```
define weak_odr dso_local i64* @__local_stdio_printf_options() local_unnamed_addr #4 {
entry:
  ret i64* @__local_stdio_printf_options._OptionsStorage
}
```

It is possible clang has a substitute `stdio.h` file it uses to compensate for Windows being insane. It's not obvious how or if it's even possible to fix this problem, or if it will cause issues with other windows header files in the future.

--- a/src/tcwrapper.cpp
+++ b/src/tcwrapper.cpp
@@ -961,9 +961,10 @@ static void optimizemodule(TerraTarget *TT, llvm::Module *M) {
 
     for (llvm::Module::iterator it = M->begin(), end = M->end(); it != end; ++it) {
         llvm::Function *fn = &*it;
-        if (fn->hasAvailableExternallyLinkage() || fn->hasLinkOnceODRLinkage()) {
+        if (fn->hasAvailableExternallyLinkage() ||
+            fn->getLinkage() == llvm::GlobalValue::LinkOnceODRLinkage) {
             fn->setLinkage(llvm::GlobalValue::WeakODRLinkage);
-        } else if (fn->hasLinkOnceLinkage()) {
+        } else if (fn->getLinkage() == llvm::GlobalValue::LinkOnceAnyLinkage) {
             fn->setLinkage(llvm::GlobalValue::WeakAnyLinkage);
         }
 #if LLVM_VERSION >= 35

--- a/src/tcwrapper.cpp
+++ b/src/tcwrapper.cpp
@@ -961,12 +961,7 @@ static void optimizemodule(TerraTarget *TT, llvm::Module *M) {
 
     for (llvm::Module::iterator it = M->begin(), end = M->end(); it != end; ++it) {
         llvm::Function *fn = &*it;
-        if (fn->hasAvailableExternallyLinkage()) {
-            fn->setLinkage(llvm::GlobalValue::WeakODRLinkage);
-        }
-#ifdef _WIN32  // On windows, the optimizer will delete LinkOnce functions that are unused
-        usedArray.push_back(fn);
-#endif
+        if (fn->isDiscardableIfUnused()) usedArray.push_back(fn);
 #if LLVM_VERSION >= 35
         if (fn->hasDLLImportStorageClass())  // clear dll import linkage because it messes
                                              // up the jit on window
@@ -1109,9 +1104,14 @@ int include_c(lua_State *L) {
 
 #ifdef _WIN32
     args.push_back("-fms-extensions");
+    args.push_back("-fms-volatile");
     args.push_back("-fms-compatibility");
     args.push_back("-fms-compatibility-version=18");
     args.push_back("-Wno-ignored-attributes");
+    args.push_back("-flto-visibility-public-std");
+    args.push_back("--dependent-lib=msvcrt");
+    args.push_back("-fdiagnostics-format");
+    args.push_back("msvc");
 #endif
 
     for (int i = 0; i < N; i++) {

--- a/src/tcwrapper.cpp
+++ b/src/tcwrapper.cpp
@@ -957,8 +957,6 @@ static void optimizemodule(TerraTarget *TT, llvm::Module *M) {
     // in some cases clang will mark stuff AvailableExternally (e.g. atoi on linux)
     // the linker will then delete it because it is not used.
     // switching it to WeakODR means that the linker will keep it even if it is not used
-    std::vector<llvm::Constant *> usedArray;
-
     for (llvm::Module::iterator it = M->begin(), end = M->end(); it != end; ++it) {
         llvm::Function *fn = &*it;
         if (fn->hasAvailableExternallyLinkage() ||

--- a/src/tcwrapper.cpp
+++ b/src/tcwrapper.cpp
@@ -961,7 +961,11 @@ static void optimizemodule(TerraTarget *TT, llvm::Module *M) {
 
     for (llvm::Module::iterator it = M->begin(), end = M->end(); it != end; ++it) {
         llvm::Function *fn = &*it;
-        if (fn->isDiscardableIfUnused()) usedArray.push_back(fn);
+        if (fn->hasAvailableExternallyLinkage() || fn->hasLinkOnceODRLinkage()) {
+            fn->setLinkage(llvm::GlobalValue::WeakODRLinkage);
+        } else if (fn->hasLinkOnceLinkage()) {
+            fn->setLinkage(llvm::GlobalValue::WeakAnyLinkage);
+        }
 #if LLVM_VERSION >= 35
         if (fn->hasDLLImportStorageClass())  // clear dll import linkage because it messes
                                              // up the jit on window
@@ -971,18 +975,6 @@ static void optimizemodule(TerraTarget *TT, llvm::Module *M) {
                                         // the jit on window
             fn->setLinkage(llvm::GlobalValue::ExternalLinkage);
 #endif
-    }
-
-    if (usedArray.size() > 0) {
-        auto *i8PtrType = llvm::Type::getInt8PtrTy(M->getContext());
-        for (auto &elem : usedArray)
-            elem = llvm::ConstantExpr::getBitCast(elem, i8PtrType);
-
-        auto *arrayType = llvm::ArrayType::get(i8PtrType, usedArray.size());
-        auto *llvmUsed = new llvm::GlobalVariable(
-                *M, arrayType, false, llvm::GlobalValue::AppendingLinkage,
-                llvm::ConstantArray::get(arrayType, usedArray), "llvm.used");
-        llvmUsed->setSection("llvm.metadata");
     }
 
     M->setTargetTriple(

--- a/src/tllvmutil.cpp
+++ b/src/tllvmutil.cpp
@@ -261,6 +261,7 @@ struct CopyConnectedComponent : public ValueMaterializer {
                 newfn = Function::Create(fn->getFunctionType(), fn->getLinkage(),
                                          fn->getName(), dest);
                 newfn->copyAttributesFrom(fn);
+                newfn->setComdat(fn->getComdat()); // copyAttributesFrom does not copy comdats
             }
             if (!fn->isDeclaration() && newfn->isDeclaration() && copyGlobal(fn, data)) {
                 for (Function::arg_iterator II = newfn->arg_begin(), I = fn->arg_begin(),

--- a/src/tllvmutil.cpp
+++ b/src/tllvmutil.cpp
@@ -261,7 +261,8 @@ struct CopyConnectedComponent : public ValueMaterializer {
                 newfn = Function::Create(fn->getFunctionType(), fn->getLinkage(),
                                          fn->getName(), dest);
                 newfn->copyAttributesFrom(fn);
-                newfn->setComdat(fn->getComdat()); // copyAttributesFrom does not copy comdats
+                newfn->setComdat(
+                        fn->getComdat());  // copyAttributesFrom does not copy comdats
             }
             if (!fn->isDeclaration() && newfn->isDeclaration() && copyGlobal(fn, data)) {
                 for (Function::arg_iterator II = newfn->arg_begin(), I = fn->arg_begin(),

--- a/tests/benchmark_fannkuchredux.t
+++ b/tests/benchmark_fannkuchredux.t
@@ -6,11 +6,10 @@
  
  ]]
 
-local C = {
-    printf = terralib.externfunction("printf", terralib.types.funcpointer(rawstring,int,true)),
-    exit = terralib.externfunction("exit", int -> {}),
-    atoi = terralib.externfunction("atoi", rawstring -> int)
-}
+C = terralib.includecstring [[
+#include<stdio.h>
+#include<stdlib.h>
+]]
 
 -- this depends highly on the platform.  It might be faster to use
 -- char type on 32-bit systems; it might be faster to use unsigned.
@@ -147,7 +146,7 @@ end))
 local N = assert(tonumber((...) or 10))
 doit(N)
 if jit.os == "Windows" then
-  terralib.saveobj("benchmark_fannkuchredux.exe", { main = main }, {"\\legacy_stdio_definitions.lib"})
+  terralib.saveobj("benchmark_fannkuchredux.exe", { main = main })
   os.execute("benchmark_fannkuchredux.exe "..N)
 else
   terralib.saveobj("benchmark_fannkuchredux", { main = main })

--- a/tests/benchmark_nbody.t
+++ b/tests/benchmark_nbody.t
@@ -2,10 +2,9 @@ local C = terralib.includecstring[[
 
 #include <math.h>
 #include <stdlib.h>
+#include <stdio.h>
 
 ]]
-
-C.printf = terralib.externfunction("printf", terralib.types.funcpointer(rawstring,int,true))
 
 pi = 3.141592653589793
 solar_mass = (4 * pi * pi)
@@ -144,7 +143,7 @@ if jit.os ~= "Windows" then
   terralib.saveobj("benchmark_nbody",{ main = main }, {"-lm"} )
   os.execute("./benchmark_nbody "..tostring(N))
 else
-  terralib.saveobj("benchmark_nbody.exe",{ main = main }, {"\\legacy_stdio_definitions.lib"} )
+  terralib.saveobj("benchmark_nbody.exe",{ main = main } )
   os.execute("benchmark_nbody.exe "..tostring(N))
 end
 

--- a/tests/run
+++ b/tests/run
@@ -6,12 +6,6 @@ local lscmd
 if ffi.os == "Windows" then
     lscmd = "cmd /c dir /b /s"
     
-    -- FIXME: https://github.com/zdevito/terra/issues/401
-    table.insert(skip, "class2.t")
-    table.insert(skip, "stdio.t")
-    table.insert(skip, "speed.t")
-    table.insert(skip, "benchmark_nbody.t")
-    table.insert(skip, "benchmark_fannkuchredux.t")
     -- Disabled due to non-deterministic failures on certain branches: https://github.com/zdevito/terra/issues/404
     table.insert(skip, "leaktest.t")
 else

--- a/tests/speed.t
+++ b/tests/speed.t
@@ -1,7 +1,7 @@
 local C = terralib.includecstring [[ 
     #include <stdlib.h>
+    #include <stdio.h>
 ]]
-C.printf = terralib.externfunction("printf", terralib.types.funcpointer(rawstring,int,true))
 
 terra doit(N : int64)
     var cur,last = 1ULL,1ULL
@@ -27,4 +27,4 @@ local test = require("test")
 print(what())
 print(test.time( function() doit:compile() end))
 print(test.time( function() doit(100000000) end))
-print(test.time( function() terralib.saveobj("speed",{main = main}, (jit.os == "Windows" and {"\\legacy_stdio_definitions.lib"} or nil)) end))
+print(test.time( function() terralib.saveobj("speed",{main = main}) end))

